CHANGELOG.md

```diff
@@ -1,3 +1,11 @@
+## v1.x.x (TBD)
+* Add support of distributed training to S3IterableDataset
+
+### Breaking changes
+* When using the S3IterableDataset with multiple workers, each worker will have access to a distinct subset 
```

Comment:
isn't it the otherway around for regular datasets. It is not sharded by default and you shard explicitly.

Reply:
`When using a [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) with [multi-process data loading](https://pytorch.org/docs/stable/data.html#multi-process-data-loading). The same dataset object is replicated on each worker process, and thus the replicas must be configured differently to avoid duplicated data. See [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) documentations for how to achieve this.`


https://pytorch.org/docs/stable/data.html#iterable-style-datasets

Reply:
Also, by not touching the constructor, you can avoid changing so many different places in the code.

Reply:
I was focused on `thus the replicas must be configured differently to avoid duplicated data` part, when was looking how others were implementing their version of IterableDataset. For example, Google implementation support multiple workers by default [dataflux_iterable_dataset](https://github.com/GoogleCloudPlatform/dataflux-pytorch/blob/main/dataflux_pytorch/dataflux_iterable_dataset.py#L169-L175). The same is true for [text_to_text](https://github.com/NVIDIA/NeMo/blob/05f75862169b1dc7f2641c54ebe5ab3f6f8451cc/nemo/collections/asr/data/text_to_text.py#L458C41-L469) from NVIDIA Nemo framework. Both of these implementations refer to PyTorch documentation on how you should develop [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset):
```
When a subclass is used with DataLoader, each item in the dataset will be yielded 
from the DataLoader iterator. When num_workers > 0, each worker process will 
have a different copy of the dataset object, so it is often desired to configure 
each copy independently to avoid having duplicate data returned from the workers.
```
It is not saying that support of sharding is mandatory, but rather that it is desired. 
Do you think it would be better to move `desired` behavior behind the lever? My intention was to make that behavior enabled by default, because one of the often reason to not use S3IterableDataset was that we duplicates data in multi worker/node environment.